### DISCUSSION TITLE
Help Themes

### DISCUSSION THREAD

**Discussion by mdonnalley** [oclif/core#871] [Ideas]
https://github.com/oclif/core/pull/852 introduces themed help output (thanks to @AllanOricil). Users can create a theme for the help output in `~/.config/<CLI>/theme.json`

In order to facilitate CLI users selecting a theme we are considering an new plugin, `@oclif/plugin-theme`, which will contain a set of commands that allow users to list, set, and unset their theme.

The big outstanding question here is how this would work best for CLI developers and their users.

**Option 1**
- CLIs must publish a `themes` directory with their CLI. `@oclif/plugin-theme` can read this directory to get a list of themes that the user can select from.
- `theme set` will take a `--theme` flag that corresponds to the file names found in the `themes` directory
- Once the user selects a theme `@oclif/plugin-theme` will copy the selected theme to `~/.config/<CLI>/theme.json`.
- At this point, the user can modify the file directly if they want

Pro: This option requires no modifications to the existing implementation other than creating the new plugin.
Con: Users are locked into having their theme at `~/.config/<CLI>/theme.json`

**Option 2**
- `theme set` will take a `--file` flag that points to _any_ json file containing a valid theme
- CLIs could publish a `themes` directory with their CLI and advise users that there are recommended themes there if they would like to use them.

Pros: Users can point the theme to any file on their machine
Cons: 
- Would require us to modify the existing implementation to not exclusively look for the theme at `~/.config/<CLI>/theme.json`
- CLIs would have less control over what users select

**Option 3**
- `theme set` will take a `--file` flag that points to _any_ json file containing a valid theme
- CLIs could publish a `themes` directory with their CLI. The themes in this directory would populate the help output for `theme set` so that users are aware of the recommended themes. `theme list` would also list out the recommended themes

Pros:
- Users can point the theme to any file on their machine
- CLIs can easily expose their recommended themes
Cons: 
- Would require us to modify the existing implementation to not exclusively look for the theme at `~/.config/<CLI>/theme.json`

### Themed `ux` methods?

Another area we'd like feedback on is whether or not the themes would be helpful outside the context of help output. For instance, themes theoretically could be applied to the `ux` methods, especially `styledObject`, `styledJSON`, `table`, and `tree`. Is that something worth implementing?

**Comment by AllanOricil** [comment]
I was thiking about asking people to publish packages to npm with keywords that can be used to identify an oclif theme. 

`keywords: ['oclif', 'theme']`


This way cli users could create and backup their theme to npm and share it with other people. 

Then there would be a `theme list` command to list all available themes using npm's search api `/v2/search?q=keywords:oclif,theme`

When running `theme install` a prompt to select a theme from a list appears and the selected theme is downloaded to `~/.config/<CLI>/`, or to a directory which all Oclif CLIs could use, the later suggestion is useful to avoid redownloading the same package multiple times for different CLIs (maybe `~/.config`?)

When running `theme install -n <package-name>` we check if it is a theme, and then install it.

When running `theme set -n <package-name>` a theme must be one of the installed packages. If the package exists in the config dir, we set it as the `~/.config/<CLI>/theme.json`

**Comment by mshanemc** [comment]
I prefer option 3, but with a tweak to solve
> Would require us to modify the existing implementation to not exclusively look for the theme at ~/.config/<CLI>/theme.json

What if `theme set --file foo.json` simply copies the contents of `foo.json` to `~/.config/<CLI>/theme.json`. (or maybe `active-theme.json` for clarity).

**Reply by mdonnalley** [reply]
@mshanemc Possibly. Although someone might be expecting to be able to modify `foo.json` directly without having to apply those changes back to `~/.config/<CLI>/theme.json`

**Comment by AllanOricil** [comment]
This is what Im thinking as the output of each command in my suggestion using npm packages:

Assuming that 3 packages with `keywords: [ 'oclif', 'theme' ]` exist in the npm registry


## theme list

`sf theme list`

step 1:
````bash
$ sf theme list
````

step 2:
````bash
Fetching available themes ...
````

step 3: if there are published themes
````bash
Available themes
- theme_a description
- theme_b description
- theme_c description
exit 0
````

step 3: when there are no published themes
````bash
sorry but we could not find themes at https://npmregistry.com
exit 0
````

## theme install

Install theme asking Users to select from a list of available themes

`sf theme install` => without `-n | --name` flag

step 1: 
````bash
$ sf theme install
````

step 2:
````bash
Fetching available themes ...
````

step 3: if there are published themes
````bash
Please, pick one and press enter
- theme_a     my precious theme a.      allanoricil
> theme_b     my precious theme b.      mshanemc
- theme_c     my precious theme c        mdonnalley
````

step 3: when there are no published themes
````bash
sorry but we could not find themes at https://npmregistry.com
exit 0
````

step 4: theme was selected
````bash
Installing `theme_b`...
`theme_b` downloaded
You can now set the theme using `sf theme set theme_b`
exit 0
````

step 4: theme failed to be downloaded
````bash
Installing `theme_b`...
`theme_b` failed to be downloaded
exit 1
````


----

Install theme using its package name

`sf theme install --name theme_c` => with `-n | --name` flag

step 1: 
````bash
$ sf theme install --name theme_c
````

step 2: 
````bash
Fetching theme_c ...
````

step 3: if theme_c exists
````bash
Installing `theme_c`...
`theme_c` downloaded
You can now set the theme using `sf theme set theme_c`
exit 0
````

step 3: if theme_c does not exist
````bash
sorry but we could not find `theme_c` at https://npmregistry.com
exit 0
````

----

Install and immediately set theme using its package name

`sf theme install --name theme_c` => with `-n | --name` and `-s|--set` flag

step 1: 
````bash
$ sf theme install --name theme_c --set
````

step 2: 
````bash
Fetching available theme_c ...
````

step 3: if theme_c exists
````bash
Installing `theme_c`...
`theme_c` downloaded
`theme_c` was set
exit 0
````

step 3: if theme_c does not exist
````bash
sorry but we could not find `theme_c` at https://npmregistry.com
exit 0
````


## theme set

Set theme placing theme.json from the package into `~/.config/<CLI>/theme.json`

`sf theme set --name theme_c` => with `-n | --name` 

step 1: 
````bash
$ sf theme set --name theme_c
````

step 2:  if theme_c exists in the host machine at SOMEWHERE (path were themes are saved)
````bash
`theme_c` was set
exit 0
````

step 3: if theme_c does not exist locally
````bash
`theme_c` was not found locally.
Please run `sf theme install --name theme_c` to install it
exit 0
````


---

## Theme Package

project structure
````bash
.
└── my-theme/
    ├── theme.json
    └── package.json
````

`package.json`
````json
{
   ...
   "keywords": [
      "oclif",
      "theme"
   ]
   ...
}
````


`theme.json`
````json
{
       "alias" : "red",
       "bin" : "rgb(255, 255, 255)",
       "command" : "#45C65A",
       "commandSummary" : "#FFFFFF",
       "dollarSign" : "#FFFF00",
       "flag" : "#45C65A",
       "flagDefaultValue" : "#1AB9FF",
       "flagOptions" : "#45C65A",
       "flagRequired" : "#FE5C4C",
       "flagSeparator" : "#FFFFFF",
       "flagType" : "#0EAEE8",
       "sectionDescription" : "#FFFFFF",
       "sectionHeader" : "#FFFF00",
       "topic" : "#45C65A",
       "version" : "#45C65A"
}
````

**Comment by cristiand391** [comment]
what if we make `~/.config/<CLI>/theme.json` store an object containing both valid themes (`themes: []`) and the current theme (`theme: gruvbox-light`), and make it the single source for themes?

I vote for option 3 + Shane's suggestion:
> What if theme set --file foo.json simply copies the contents of foo.json to ~/.config/<CLI>/theme.json. (or maybe active-theme.json for clarity).

then for `theme set --file foo.json` to work:
 1. `foo.json` is expected to be a json file containing **just 1 theme** [^1]
 2. push the theme in `foo.json` to  `~/.config/<CLI>/theme.json`.themes array
 3. set it in the `theme` prop

[^1]: we should enforce a theme to have a name (`themeName` prop?)


### use-case:
I plan to have 1 light and 1 dark theme, then switch it in a script that [runs automatically when switching to light/dark mode](https://arslan.io/2021/02/15/automatic-dark-mode-for-terminal-applications/).
I want to keep both in the same file and set one by updating the `theme` prop in the json via `jq` or `theme set --theme` and would like to avoid all the copy/paste on-the-fly.


### CLIs shipping themes

CLIs can ship a `themes` folder containing a single json file that have themes (in an array), if you want to implement a `theme list` command it could follow this order:
1) read themes in ``~/.config/<CLI>/theme.json``
2) read themes in CLI's themes folder
3) print those by name


### Sharing oclif themes
I'm not sure how many people would use this feature, let alone start publishing and maintain an npm package.
IMO, we can leave this up to users sharing either a repo on GH **and/or** keep a ["theme gallery"](https://zellij.dev/documentation/theme-gallery) on `oclif.io` (users would be able to put theirs there). 

if requested, we could make `theme set` accept themes from a URL:
`sf theme set --url https://raw.githubusercontent.com/cristiand391/oclif-gruvbox/main/gruvbox-light.json`


See also:
* how zellij handles themes: https://zellij.dev/documentation/themes
* zellij themes on GH: https://github.com/search?q=zellij%20themes&type=repositories
* themes in alacritty (similar to zellij approach): https://github.com/fatih/dotfiles/blob/52e459c991e1fa8125fb28d4930f13244afecd17/alacritty.yml#L94
* alacritty themes on GH: https://github.com/search?utf8=%E2%9C%93&q=alacritty+theme&type=repositories

**Reply by mdonnalley** [reply]
Love the idea of having a single file containing a `theme` prop and a `themes` array

But I'm stuck on wanting people to continue to edit the file after setting it.

I'm imagining someone running `theme set --file foo.json` and then expecting to be able to continue to make edits to that file and have those changes immediately reflected. I don't think it would be obvious to the user that the command really just copies the contents of their file to another and they need to run `theme set` again after making changes... I guess we could call the command `theme add` to make that more obvious?

Perhaps it should work something like this:
- `~/.config/<CLI>/theme.json` would look like this: `{theme: 'dark', themes: [{name: 'dark', ...}]}`
- `theme list` would list all the themes under `themes` and any themes shipped with the CLI
- `theme set` would only have a `--name` flag that corresponds to the `themes` 
- users can directly modify `~/.config/<CLI>/theme.json` if they choose
- or they can use `theme add --file` or `theme add --url` to insert a new theme

**Reply by cristiand391** [reply]
> I guess we could call the command theme add to make that more obvious?

💯 

I like your proposal, setting the theme via `jq` will be faster for scripting and `theme set|add` will work for most cases 👍🏼

**Reply by mdonnalley** [reply]
@AllanOricil had a [good point](https://github.com/oclif/plugin-theme/pull/7#issuecomment-1815321276) that people might want to reuse themes across oclif CLIs (e.g. same theme for both `sf` and `heroku`). The previous proposal would require users to duplicate that theme across multiple directories

With that in mind, here's another proposal based on Allan's comment:
- `~/.config/<CLI>/theme.json` contains a single theme, which is the active theme for the CLI
- CLIs can ship a `themes` folder containing a variety of recommended themes (e.g. `dark.json`, `light.json`)
- `theme list` will list the themes under the published `themes` folder
  - Important to note that this will **not** list any of the user's custom themes - maybe we should call it `theme recommended list` or something like that??
- `theme set [NAME]` will copy the contents of a published theme to `~/.config/<CLI>/theme.json`
- `theme link --file my-theme.json` will create a symlink between `~/.config/<CLI>/theme.json` and the provided file.
  - This allows users to easily reuse themes across oclif CLIs and allows them to continue making edits to the file without having to rerun `theme set`
  - @cristiand391 I think this would also support your use case of automatically switching themes? You would just need to call `theme link --file my-light-theme.json` or `theme link --file my-dark-theme.json` based on whatever context you're working from

And if there's demand for sharing themes we can add `theme download --url`, where the the `--url` flag is a `https://raw.githubusercontent.com/` url (or any url that points to a json file) 

Thoughts?

**Reply by cristiand391** [reply]
Yes, that would work for my use case, but I'm personally not concerned about having duplicates for +1 oclif CLIs.

Thinking about usage of this feature:

In my case I'll switch between a dark and light theme a few times per day (script runs when switching light/dark mode on mac), but other than that I don't expect myself to add new or update themes frequently so even If I would like to share a theme in +1 oclif CLIs I wouldn't mind having duplicates because:

1) I don't expect a theme to be upgraded frequently (I'll publish a gruvbox light/dark one and probably call it "finished", only updating it if oclif makes a breaking change)
2) themes aren't big or distributed in multiple files.
3) Even if you have like +5 oclif CLIs, you set themes just 1 time, if you worry about those being not synced you can script that. 

about 3: this feels like a very specific case that we can safely support without exposing FS stuff via `theme link` if we keep a themes array and the current in one file.

I feel confident saying that most people don't automate dark/light themes for terminal stuff, in those cases users will just use 1 theme that looks good on light/dark mode... so other than updating or trying a new theme every now and then, your local theme will not change that frequently to have to worry about your other's CLI theme being stale/duplicated.
